### PR TITLE
feat(aio): revise sitemap.service with sitemap.json

### PR DESCRIPTION
**Pull request by wardbell**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```
Replaces #14379

This PR adds "sitemap" metadata based on a realistic variety of navigation hierarchy requirements embodied in a sample version of `sitemap.json`. The real version would be hand-edited by docs team.

This PR explores, through unit tests, how consumers (e.g. NavEngine, the navigation panel, the menu bar) can get docs from metadata and get metadata from a path. The latter capability will be tied into the aio router.

Todo:
- [ ] Confirm terminology
- [ ] Confirm content location and adjust

### CODE REVIEW

**Comment by wardbell** [comment]
## Why author-edited `sitemap.json`

@petebacondarwin has asked on various occasions about generating `sitemap.json` with dgeni rather than having it as a separate file to be manually edited.

I feel strongly that this is not a best use of our efforts at this time. Here are a few of my many reasons:

1. We don't know what the schema should be yet. We haven't even put it to work yet. Not a good idea to write a generator for something that is so provisional. 

1. There are only a few docs to worry about (<100). We could easily create the entire `sitemap.json` in under three hours and we'd be done.  There is no chance that we could create a `sitemap.json` generator AND populate the docs with the source information in anything close to that time.

1. Given our short time to delivery, we should work on something critical path. We don't need a generated sitemap to release the new DocViewer. You have much more important and helpful contributions to make.

1. Generating the `sitemap.json` doesn't solve a problem that we actually have. Today's authors are quite satisfied with editing a separate navigation json file. What complicates it today is that _there are too many_ of those damned `_data.json` files, scattered in fixed locations all over the place, each location associated with an unhelpful information architecture.  Having _just one JSON file_ to edit will be a dream!

1. It's not clear that it will _ever_ be a good idea to generate the navigation out of the docs.

    1. There's no good reason to believe that a document should know where it belongs in the navigation.
    1. It means any re-org of navigation involves touching the affected documents. Even a minor re-org could easily touch 20+ docs.  And we know for certain that the navigation will change several times over the next few months.
    1. A re-org PR looks like a document editing PR.
    1. Many navigation items have no corresponding docs. Section headers (containers of sub-docs). Links to external pages (`Gitter`, `StackOverflow`).
    1. Editing the navigation means editing docs and seeing what dgeni generates.
    1. Can't re-org w/o re-running tooling. This hampers Jules.
    1. Coping with all of these objections will like lead to having _some_ nav info in the docs and _other_ nav info in a _separate JSON file_. That's a horror show.

1. If we change our mind later, we can use the handcrafted `sitemap.json` to populate the docs automatically and roll from there.

I think I've clubbed this (dead?) horse enough ;-)

File: aio/src/content/api/common/date-pipe.html [hunk 1]
```diff
@@ -0,0 +1,239 @@
+<header class="hero background-sky"><h1 class="hero-title is-standard-case">DatePipe</h1><span
```
Comment (IgorMinar) [inline aio/src/content/api/common/date-pipe.html]: why is this file here?

dgeni now generates html files for us from the source code

File: aio/src/content/guide/overview.html [hunk 2]
```diff
@@ -0,0 +1 @@
+TODO: guide overview doc
```
Comment (IgorMinar) [inline aio/src/content/guide/overview.html]: this file should not be here. we should have an `overview.md` file in `aio/content/guide/overview.md`. no? is this just a mock?

File: aio/src/content/sitemap.json [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    "path": "guide/quickstart",
+    "primary": true,
+    "title": "Quickstart",
+    "tooltip": "A quick look at an Angular app."
+  },
+
+  "cli-quickstart": {
+    "path": "guide/cli-quickstart",
+    "title": "CLI Quickstart",
+    "tooltip": "A quick look at an Angular app built with the Angular CLI."
```
Comment (IgorMinar) [inline aio/src/content/sitemap.json]: I'm very uncomfortable having the metadata in this file. Why not put them into the md file using some dgeni-recognizable syntax? that way metadata is together with the content (just like metadata decorators keep class metadata together with the class).

Comment (IgorMinar) [inline aio/src/content/sitemap.json]: the way I understood the requirements this file should just contain the menu structure and list of documentIDs that belong to a given section:
```json
{
  "top-menu": ["home", "features", "news"],
  "side-menu": [
    { "section-name": "core documentation",
      "items": [
        "quickstart",
        "cli quickstart",
        { "group": "guide",
           "items": ["guide/overview", "guide/setup", "..."]
         }
  ] 
```

this way we have all the flexibility to create synthetic grouping of docs and nest them as much as we want, but all the document metadata remains in each document.

File: aio/karma.conf.js [hunk 4]
```diff
@@ -30,8 +34,8 @@ module.exports = function (config) {
       environment: 'dev'
     },
     reporters: config.angularCli && config.angularCli.codeCoverage
-              ? ['progress', 'karma-remap-istanbul']
-              : ['progress'],
+              ? ['progress', 'karma-remap-istanbul'] //, 'kjhtml']
+              : ['progress', 'kjhtml'],
```
Comment (IgorMinar) [inline aio/karma.conf.js]: this karma change is unrelated to the sitemap feature. please split it into a separate commit with type `build(aio):`.

it's a good practice to always split code changes into commits based on the logical grouping. multiple, well documented commits are easier to review as well as cherry-pick or if needed revert.

File: aio/src/app/doc-viewer/doc-viewer.component.ts [hunk 5]
```diff
@@ -105,7 +105,7 @@ export class DocViewerComponent implements DoCheck, OnDestroy {
 
 class DisplayedDoc {
 
-  metadata: DocMetadata;
+  metadata: NavigationMapEntry;
```
Comment (IgorMinar) [inline aio/src/app/doc-viewer/doc-viewer.component.ts]: I think this a wrong change - as I mentioned in the other comment the metadata should be part of the doc and not something specific to navigation. it's not like the title of the doc changes based on how we navigated to it.

File: aio/src/app/nav-engine/doc.model.ts [hunk 6]
```diff
@@ -1,10 +1,67 @@
...
   content: string;
 }
+
+/**
+ * Entry in the UI NavigationMap that describes a document or container of documents
+ * Each entry corresponds to a link in the UI.
+ * An entry has either a path or a url. If it has neither, it's a pure container.
+ * An entry with entries is a container.
+ * An entry with a path or a url is a document
+ * An entry with both a path and entries is both a container and a document.
```
Comment (IgorMinar) [inline aio/src/app/nav-engine/doc.model.ts]: the verbosity and complexity of this jsdoc tells me that we are trying to model the data incorrectly - too much responsibility in one place, overloaded meaning of a single interface.

File: aio/src/app/nav-engine/doc.model.ts [hunk 7]
```diff
@@ -1,10 +1,67 @@
...
+  /** If defined, this entry is a container of these child entries */
+  entries?: NavigationMapEntry[];
+}
+
+/**
+ * The navigation hierarchy for the site.
+ * Immediate properties are top-level sections.
+ */
+export interface NavigationMap {
+  [index: string]: NavigationMapEntry;
```
Comment (IgorMinar) [inline aio/src/app/nav-engine/doc.model.ts]: why not just have a `Map<string, NavigationMapEntry>`? that's succinct enough and we don't need a custom type that everyone needs to get familiar with.

(I'm not sure if this interface is needed at all given the comments above)

**Review by IgorMinar — CHANGES_REQUESTED** [review]

**Comment by IgorMinar** [comment]
wrt @petebacondarwin suggestion to generate the sitemap that @wardbell vehemently opposed in the comment above: I think that a good compromise is to redo the sitemap using the minimalistic approach I suggested above:

- keep metadata in the doc and use dgeni to extract it and make it available to the viewer
- (at least for now) do not put information about the menu placement into the doc metadata but keep it in the minimalistic sitemap doc

There is a risk however - it's possible to create an md file that is not reachable from anywhere. To mitigate this risk, we can have dgeni read in the sitemap and check it against the list of available docs and ensure that all of them are reachable from somewhere (it might get more complicated if some docs are only reachable from a link of a doc that is listed in the menu, but since dgeni knows about all the links between docs, it could handle that as well).

File: aio/src/content/api/common/date-pipe.html (hunk 1 above)
Comment (wardbell) [inline aio/src/content/api/common/date-pipe.html]: It's temporary ... just like _every other file_ in the `content` directory including the marketing files. 

I did not know that there is a way to generate these files. I do know that, without them, the e2e tests would fail and, AFAIK, we have not (yet) made file generation part of our CI build process.

I'll be thrilled when all of these forces are in place. When we're ready, there will be no content in the `content` folder.

File: aio/src/content/guide/overview.html (hunk 2 above)
Comment (wardbell) [inline aio/src/content/guide/overview.html]: It's garbage. It was there before and I didn't know why but didn't want to discard it until I did know why. Happy to delete it.

File: aio/src/content/sitemap.json (hunk 3 above)
Comment (wardbell) [inline aio/src/content/sitemap.json]: Let's discuss. My own experience of with the docs tells me that none of these data belong in the docs. Everything here is about presentation in the UI ... including the title, tooltip, etc. 

Dividing presentation data into two parts, some in the docs, some here, i just going to make this harder to maintain.

But my best argument is that we shouldn't be messing with this now. We are going to waste days of effort figuring this out, modifying docs so that the titles and tooltips and all that stuff appear the way they should (which no one knows right now anyway).

We need this spend this time doing _real work_.  I put this together based on the information we use now in the `_data.json` files.  That should suffice for now. We can come back after we ship and revisit.  But if we invest in generating this now, we will not have time to do the critical pieces ... such as generating the HTML content from today's Jade, making search work, getting the navigation up and running, making the viewer presentable and ensuring it performs well.

Much more discussion on this point is bike shedding of the worst sort. I'd happily surrender on every aspect of this ... if only someone had an alternative that wasn't going to take 10x as long to implement and maintain. I don't believe that alternative exists.

File: aio/src/app/nav-engine/doc.model.ts (hunk 6 above)
Comment (wardbell) [inline aio/src/app/nav-engine/doc.model.ts]: Every single item in this model comes from our docs today. I'm not making this up.

File: aio/src/app/nav-engine/doc.model.ts (hunk 7 above)
Comment (wardbell) [inline aio/src/app/nav-engine/doc.model.ts]: Works for me.

File: aio/karma.conf.js (hunk 4 above)
Comment (wardbell) [inline aio/karma.conf.js]: I'll split this out into a separate PR.

**Comment by petebacondarwin** [comment]
Oh dear I think I was misunderstood. I never wanted to auto generate the site map data. I was talking about the meta data file

**Comment by wardbell** [comment]
My written words are perhaps more argumentative than I intended. Sorry! I don't care _nearly as much_ as they seem. I just don't want to get trapped into generation and doc editing tasks when we're short on time.  We should be able to power thru this in the morning.

I've re-freshed this PR (with a **push -f**) with changes recommended above _except_ I haven't changed the `sitemap.json` or the the `NavigationNode` (fka `NavigationMapEntry`) ... yet.

Please note about the `NavigationNode`  ... we'll need that info _within the app_, no matter how it is derived, in order to make the multi-level left nav behave. I'm happy to explain how each of its properties are used.

File: aio/src/app/doc-viewer/doc-viewer.component.ts (hunk 5 above)
Comment (wardbell) [inline aio/src/app/doc-viewer/doc-viewer.component.ts]: Let's not confuse where these data come from with what we need associated with the doc to make the navigation UX work. I've renamed `metadata` to `node` (and `NavigationMapEntry` to `NavigationNode`) to make it clearer that we're keeping the doc and its node together.

**Comment by wardbell** [comment]
Closed by #14429
